src/Microsoft.ApiDesignGuidelines.Analyzers/Core/ImplementIDisposableCorrectly.cs

```diff
@@ -106,13 +112,494 @@ public abstract class ImplementIDisposableCorrectlyAnalyzer : DiagnosticAnalyzer
+            /// </summary>
+            private static void CheckProvideDisposeBoolRule(INamedTypeSymbol type, SymbolAnalysisContext context)
+            {
+                context.ReportDiagnostic(type.CreateDiagnostic(ProvideDisposeBoolRule, type.Name));
+            }
+
+            /// <summary>
+            /// Checks rule: Ensure that {0} is declared as protected, virtual, and unsealed.
+            /// </summary>
+            private static void CheckDisposeBoolSignatureRule(IMethodSymbol method, INamedTypeSymbol type, SymbolAnalysisContext context)
+            {
+                if (method.DeclaredAccessibility != Accessibility.Protected ||
+                    !(method.IsVirtual || method.IsAbstract || method.IsOverride) || method.IsSealed)
+                {
+                    context.ReportDiagnostic(method.CreateDiagnostic(DisposeBoolSignatureRule, $"{type.Name}.{method.Name}"));
+                }
+            }
+
+            /// <summary>
+            /// Checks rule: Modify {0} so that it calls Dispose(true), then calls GC.SuppressFinalize on the current object instance ('this' or 'Me' in Visual Basic), and then returns.
+            /// </summary>
+            private void CheckDisposeImplementationRule(IMethodSymbol method, INamedTypeSymbol type, ImmutableArray<IOperation> operationBlocks, OperationBlockAnalysisContext context)
+            {
+                var validator = new DisposeImplementationValidator(suppressFinalizeMethod, type);
+                if (!validator.Validate(operationBlocks))
+                {
+                    context.ReportDiagnostic(method.CreateDiagnostic(DisposeImplementationRule, $"{type.Name}.{method.Name}"));
+                }
+            }
+
+            /// <summary>
+            /// Checks rule: Modify {0} so that it calls Dispose(false) and then returns.
+            /// </summary>
+            private static void CheckFinalizeImplementationRule(IMethodSymbol method, INamedTypeSymbol type, ImmutableArray<IOperation> operationBlocks, OperationBlockAnalysisContext context)
+            {
+                // TODO: Implement check of Finalize
+            }
+
+            /// <summary>
+            /// Checks if type implements IDisposable interface or an interface inherited from IDisposable.
+            /// Only direct implementation is taken into account, implementation in base type is ignored.
+            /// </summary>
+            private bool ImplementsDisposableDirectly(ITypeSymbol type)
+            {
+                return type.Interfaces.Any(i => i.Inherits(disposableType));
+            }
+
+            /// <summary>
+            /// Checks if base type implements IDisposable interface directly or indirectly.
+            /// </summary>
+            private bool ImplementsDisposableInBaseType(ITypeSymbol type)
```

Comment:
`ImplementsDisposableInBaseType` returns true, if any base type in inheritance chain implements `IDisposable`. For example, when analyzing class `D : IDisposable`, a diagnostic must be reported when following is true:
- D inherits from C
- C inherits from B
- B inherits from A
- A implements IDisposable

Then D re-implements IDisposable from A.

Although the method does not need to explicitly recurs through all base types. `AllInterfaces` does that.
Added test cases `CSharp_CA1063_IDisposableReimplementation_Diagnostic_ReimplementingIDisposableWithDeepInheritance` and `Basic_CA1063_IDisposableReimplementation_Diagnostic_ReimplementingIDisposableWithDeepInheritance` in 3a411e0cab54cc103643195f2341ee8e14b4630f